Brad/graphs+bubble sort+merge sort+quicksort

benstone1 opened the pull request.

Evilsmaher committed:
Graphs+Bubble+Merge

Adding Playgrounds for:

- Graphs
- Bubble Sort
- Merge Sort

Bubble Sort and Merge Sort are paired with examples.

Graphs Playground has the final codebase at the top with the individual exercises and associated code in sections below with numbers. xa is the pre-code and xb is the solution. For example, 1a is the code the student's will initially see and 1b is the solution code for that exercise.

Evilsmaher committed:
Merge pull request #3 from Evilsmaher/main

Graphs+Bubble+Merge

Evilsmaher committed:
Graphs + Mergesort Updates + Quicksort Implementation(s)

Mergesort:
- The code is a bit longer with the following modifications
-- removed `removeFirst()`
-- I use "pointers" to track adding elements in the correct order

- Graphs:
-- Equatable comparisons use `===` comparing instances vs instance data
-- Changed `graphEdge` to a structure instead of class
-- Changed `bidirectional` to `isBidirection`

- Quicksort:
-- Added 2 quicksort implementations to look over.
--- The first implementation is the most common online but doesn't truly capture the advantages that quicksort provides
--- The second implementation is a bit more advanced but has the right idea and run time is more efficient.

Evilsmaher committed:
Merge pull request #4 from Evilsmaher/main

Graphs + Mergesort Updates + Quicksort Implementation(s)

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+  guard array.count > 1 else { return array }
+
+  let pivot = array[array.count/2]
+  let less = array.filter { $0 < pivot }
```
`less` -> `smallerElements`

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+
+  let pivot = array[array.count/2]
+  let less = array.filter { $0 < pivot }
+  let equal = array.filter { $0 == pivot }
```
`equal` -> `equalElements`

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+  let pivot = array[array.count/2]
+  let less = array.filter { $0 < pivot }
+  let equal = array.filter { $0 == pivot }
+  let greater = array.filter { $0 > pivot }
```
`greater` -> `greaterElements`

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+  let equal = array.filter { $0 == pivot }
+  let greater = array.filter { $0 > pivot }
+
+  return quicksort1(less) + equal + quicksort1(greater)
```
Update this to match the new names

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+ */
+// 1.
+
+func quicksort1(_ array: [Int]) -> [Int] {
```
This implementation looks great!  The higher-order filters are very clean.

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+func quicksort1(_ array: [Int]) -> [Int] {
+  guard array.count > 1 else { return array }
+
+  let pivot = array[array.count/2]
```
Add a space before and after the `/` operator.

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+func partition(_ array: inout [Int], from start: Int, to end: Int) -> Int {
+  let pivot = array[end]
+  
+  var lessThanPivotIncrementer = start
```
LMK: I think `currentIndex` might be a better name for this variable.

Evilsmaher replied:
I think `currentValue` would be better since it is the actual value. This also plays nicely into the statement `if array[value] <= currentValue` vs `if array[value] <= currentIndex`.

benstone1 commented on Track5/Quick Sort/Quick Sort.playground/Contents.swift:
```diff
@@ -0,0 +1,51 @@
+// 2.
+func quicksort2(_ array: inout [Int], from start: Int, to end: Int) {
+  if start < end {
+    let partititon = partition(&array, from: start, to: end)
```
The variable `partition` should have a different name than the method.  Maybe something like `partitionIndex`?

Evilsmaher committed:
Update Mergesort

- Condense mergesort to use `dropFirst(_:)` instead of iterating the arrays
- Removed the final `else` and replaced it with the statement above.

Evilsmaher committed:
Merge pull request #7 from Evilsmaher/main

Update Mergesort

Evilsmaher committed:
Update Quicksort

- Made the changes identified in Github
- Also changed `value` to `index`

Evilsmaher committed:
Merge pull request #8 from Evilsmaher/main

Update Quicksort

Evilsmaher committed:
Graphs + Graphs Search

Graphs:
- Changed name from "Graph Search" to "Graphs

Graph Search:
- Implemented BFS + DFS in 5 Stages
1: DFS for exploring nodes from a start to end node for acyclic graphs
2: DFS for exploring nodes from a start to end node allowing for cyclic graphs
3: DFS for exploring all nodes from a node to all nodes in order of DFS
4: BFS for exploring nodes from a start to end node for acyclic and cyclic graphs
5: BFS for exploring all nodes from a start node

Evilsmaher committed:
Merge pull request #9 from Evilsmaher/main

Graphs + Graphs Search

Evilsmaher committed:
Updated Graphs Lesson

- Changed all `addEdge(_:_:)` to `addEdge(from:to:)` for each exercise.
- Modified the bidirectional support to be added when undirected graphs are shown instead of acyclic graphs. This seemed to be more logical.
- Changed the exercise for acyclic graphs and annotated how it is different than trees.
- Changed the exercise for cyclic graphs and annotated how it differs from acyclic.
- Changed `GraphEdge` initializer for weight to be an `Int?` instead of `Int`. Since weight is optional in the method call, as well as an optional instance variable, this makes the most sense instead of checking if it is `nil` or not.
- Added an overloaded function `addEdge(from:to:)` that allows for weights and one that does not have weights. This implementation makes it where you don't have to put `(NODE, BOOL, nil)` the `nil` value in everytime for a graph that does not have edges. This matches the idea that weights are optional.
- Added an exercise for the Review (8) exercise.

Evilsmaher committed:
Merge pull request #10 from Evilsmaher/main

Updated Graphs Lesson
